torch/distributed/_tensor/experimental/attention.py

```diff
@@ -634,27 +724,136 @@ def backward_hook(grad: torch.Tensor) -> None:
         return tuple(out)
 
 
-class _CausalBehavior(Enum):
-    SKIP = None
-    NOT_IS_CAUSAL = False
-    IS_CAUSAL = True
+@contextlib.contextmanager
+def _context_parallel(seq_dim: int, mesh: DeviceMesh) -> Generator[None, None, None]:
+    """Replace SDPA with the CP-wrapped version and enable DTensor CP dispatcher."""
 
+    def attention_input_fn(
+        mesh: DeviceMesh, *args: Tuple[Any, ...], **kwargs: Dict[str, Any]
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        placement = [Shard(seq_dim)]
+        all_args = []
 
-def _is_causal_behavior(
-    rank: int, world_size: int, i: int, is_causal: bool
-) -> _CausalBehavior:
+        for arg in itertools.chain(args, kwargs.values()):
+            if isinstance(arg, torch.Tensor) and not isinstance(arg, DTensor):
+                arg = DTensor.from_local(arg, mesh, placement, run_check=False)
+
+            all_args.append(arg)
+
+        new_args = tuple(all_args[0 : len(args)])
+        new_kwargs = dict(zip(kwargs.keys(), all_args[len(args) :]))
+        return new_args, new_kwargs
+
+    def attention_output_fn(mesh: DeviceMesh, outputs: Any) -> Any:
+        new_outputs = []
+        for output in [outputs] if isinstance(outputs, torch.Tensor) else outputs:
+            output = output.to_local() if isinstance(output, DTensor) else output
```

Comment:
for my education, what is our assumption about user-original inputs/outputs before `context_parallel` wrapper gets involved?

it looks a little asymetric:
  - input_fn says "if args are not DTensor, make them DTensor sharded on seq_dim"
   - output_fn says "if output is a DTensor, unpack its local part
But how does output_fn know if user already provided a DTensor as input and expects a DTensor back at output vs a local part?

Reply:
Yes, so this part currently only support FSDP + CP and also with TorchTitan TP because in TorchTitan we convert it to local tensor before calling SDPA. But for completeness, this must be handled correctly in the `attention_output_fn`, so we can support the general case where we don't convert the DTensor (TP/SP) to the local tensor. That one I'm still working on.